### PR TITLE
parallel graph match

### LINKED ISSUE #300: Parallelize FAQ random initializations

### PR DESCRIPTION
**Pull request by SHAAAAN**
Closes #300 

#### What does this implement/fix? Explain your changes.

This PR includes parallelized graph match algorithm.

#### Any other comments?

The outputs of comparison between not parallelized and parallelized algorithms for running time and match ratio are as follows. 

Parameters:
vertices: 99*99
initial times: 30
gmp: True
init_method: "rand"
seed number: [3, 8]
runs: 30 graphs / seed number

Screen Shot 2020-05-02 at 4 47 11 PM

Screen Shot 2020-05-02 at 4 47 17 PM

#### Reference Issues/PRs

Reference
[1] J.T. Vogelstein, J.M. Conroy, V. Lyzinski, L.J. Podrazik, S.G. Kratzer,
        E.T. Harley, D.E. Fishkind, R.J. Vogelstein, and C.E. Priebe, “Fast
        approximate quadratic programming for graph matching,” PLOS one, vol. 10,
        no. 4, p. e0121002, 2015.
[2] D. Fishkind, S. Adali, H. Patsolic, L. Meng, D. Singh, V. Lyzinski, C. Priebe,
        Seeded graph matching, Pattern Recognit. 87 (2019) 203–215

### CODE REVIEW

File: graspy/match/gmp.py [hunk 1]
```diff
@@ -280,7 +274,7 @@ def f(x):  # computing the original optimization function
                 n_iter += 1
             # end of FW optimization loop
 
-            row, col = linear_sum_assignment(
+            _, col = linear_sum_assignment(
```
Comment (bdpedigo) [inline graspy/match/gmp.py:277]: why this change?

File: graspy/match/gmp.py [hunk 2]
```diff
@@ -226,7 +220,7 @@ def fit(self, A, B, seeds_A=[], seeds_B=[]):
         B21T = np.transpose(B21)
         B22T = np.transpose(B22)
 
-        for i in range(self.n_init):
+        def forloop(init_num):
```
Comment (bdpedigo) [inline graspy/match/gmp.py:223]: can you think of a different name for this, maybe _fit_single_init() or something like that

File: graspy/match/gmp.py [hunk 3]
```diff
@@ -18,22 +18,21 @@
 from scipy.optimize import minimize_scalar
 from sklearn.utils import check_array
 from sklearn.utils import column_or_1d
-from .skp import SinkhornKnopp
+from graspy.match.skp import SinkhornKnopp
```
Comment (bdpedigo) [inline graspy/match/gmp.py:21]: shouldn't need to change this

Comment (bdpedigo) [reply graspy/match/gmp.py:21]: can you explain why

File: graspy/match/gmp.py (hunk 1 above)
Comment (SHAAAAN) [reply graspy/match/gmp.py:277]: Row is not used in this algorithm, so I didn't store them. 
If you think we potentially need this later, I can add it back.

File: graspy/match/gmp.py (hunk 3 above)
Comment (SHAAAAN) [reply graspy/match/gmp.py:21]: I wrote this file in a different folder so I couldn't use ".skp". I'll change it back.

File: graspy/match/gmp.py (hunk 2 above)
Comment (SHAAAAN) [reply graspy/match/gmp.py:223]: Sure.

File: graspy/match/gmp.py [hunk 4]
```diff
@@ -226,7 +220,7 @@ def fit(self, A, B, seeds_A=[], seeds_B=[]):
         B21T = np.transpose(B21)
         B22T = np.transpose(B22)
 
-        for i in range(self.n_init):
+        def _fit_single_init(init_num):
```
Comment (bdpedigo) [inline graspy/match/gmp.py:223]: init_num is not used, correct? if so, remove as arg and don't pass anything to this function

File: graspy/match/gmp.py [hunk 5]
```diff
@@ -290,11 +284,29 @@ def f(x):  # computing the original optimization function
...
-
-            if obj_func_scalar * score_new < obj_func_scalar * score:  # minimizing
-                score = score_new
-                perm_inds = np.zeros(n, dtype=int)
-                perm_inds[permutation_A] = permutation_B[perm_inds_new]
+            return score_new, perm_inds_new
+            # end of the _fit_single_init function
+
+        par = Parallel(n_jobs=-1)
+        score_new, perm_inds_new = map(
```
Comment (bdpedigo) [inline graspy/match/gmp.py:291]: this is too much for one line. can you have `outs=par(...)` on a separate line or something like that

**Comment by SHAAAAN** [comment]
> also need at least one test showing that this works with n_jobs > 1. ideally, take one of the tests that already exists and just make it run in parallel, maybe just 2 inits

Hi Pedigo, for this issue, can I add parameters of "n_jobs=-1" and "par=True" for the __init__ part in gmp.py? And then I can set "par=False" and "n_jobs = 5" for the test?

File: graspy/match/gmp.py (hunk 3 above)
Comment (bdpedigo) [reply graspy/match/gmp.py:23]: can you fix this

File: graspy/match/gmp.py [hunk 6]
```diff
@@ -101,6 +102,7 @@ def __init__(
         shuffle_input=True,
         eps=0.1,
         gmp=True,
+        n_jobs=-1,
```
Comment (bdpedigo) [inline graspy/match/gmp.py:105]: 1

File: graspy/match/gmp.py [hunk 7]
```diff
@@ -136,6 +138,11 @@ def __init__(
         else:
             msg = '"gmp" must be a boolean'
             raise TypeError(msg)
+        if (n_jobs > 0 and type(n_jobs) is int) or n_jobs == -1:
```
Comment (bdpedigo) [inline graspy/match/gmp.py:141]: remove this stuff, n_jobs can be -2, -3, etc. I think I am fine with just letting joblib to the input checking

Comment (SHAAAAN) [reply graspy/match/gmp.py:141]: > remove this stuff, n_jobs can be -2, -3, etc. I think I am fine with just letting joblib to the input checking

Sure

**Comment by SHAAAAN** [comment]
> Thanks Shan, looks done to me! I have some other changes I'm making to GMP so I may wait a bit to merge but you are all good

Thanks for your help along the way!

**Comment by bdpedigo** [comment]
closing for now because we are going to be refactoring the internals of `GraphMatch`, right @asaadeldin11? But may be able to use some of this to make that code parallel once we do?
